File: Dsa-day59.java
class Solution {
    public String[] spellchecker(String[] wordlist, String[] queries) {
        Set<String> exact = new HashSet<>();
        Map<String,String> cap = new HashMap<>();
        Map<String,String> vow = new HashMap<>();
        for (String w : wordlist) exact.add(w);
        for (String w : wordlist) {
            String lw = w.toLowerCase();
            cap.putIfAbsent(lw, w);
            vow.putIfAbsent(devowel(lw), w);
        }
        String[] res = new String[queries.length];
        for (int i=0;i<queries.length;i++) {
            String q = queries[i];
            if (exact.contains(q)) {res[i]=q; continue;}
            String lq = q.toLowerCase();
            if (cap.containsKey(lq)) {res[i]=cap.get(lq); continue;}
            String vq = devowel(lq);
            res[i] = vow.getOrDefault(vq, "");
        }
        return res;
    }
    private String devowel(String s) {
        StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray()) {
            if ("aeiou".indexOf(c)>=0) sb.append('*');
            else sb.append(c);
        }
        return sb.toString();
    }
}

File: Dsa-day59.cpp
class Solution {
public:
    vector<string> spellchecker(vector<string>& wordlist, vector<string>& queries) {
        unordered_set<string> exact(wordlist.begin(), wordlist.end());
        unordered_map<string,string> cap;
        unordered_map<string,string> vow;
        auto devowel = [](string s) {
            for (auto &c : s) if (string("aeiou").find(c)!=string::npos) c='*';
            return s;
        };
        for (auto &w : wordlist) {
            string lw=w;
            for (auto &c:lw) c=tolower(c);
            if (!cap.count(lw)) cap[lw]=w;
            string vw=devowel(lw);
            if (!vow.count(vw)) vow[vw]=w;
        }
        vector<string> res;
        for (auto &q : queries) {
            if (exact.count(q)) {res.push_back(q); continue;}
            string lq=q;
            for (auto &c:lq) c=tolower(c);
            if (cap.count(lq)) {res.push_back(cap[lq]); continue;}
            string vq=devowel(lq);
            if (vow.count(vq)) res.push_back(vow[vq]);
            else res.push_back("");
        }
        return res;
    }
};
